### ISSUE TITLE
model save and validation

### ISSUE THREAD

**Issue by will123195** [SkyPHP/skyphp#139]
Let's get rid of model `set_` methods.  

This is what I'm thinking for `save()`:
- `beginTransaction()`
- `runValidation()`
  1. **`beforeCheckRequiredFields()`**
  2. `checkRequiredFields()`
  3. **`validate()`**
- **`beforeInsert()` or `beforeUpdate()`**
- `saveArray()`
- **`beforeRefreshCache()`**
- `reload(true)`
- **`afterInsert()` or `afterUpdate()`**
- `refreshBelongsTo()`
- `commitTransaction()` or `rollbackTransaction()`
- `return $this` or `throw new Exception`

Also, 

`saveArray()` needs to be private so you can't bypass validation, needs better documentation (pass through ids?), and should set `$this->id` after insert.

I think `reload()` should have just one bool param `$refresh_cache`.

**Comment by will123195** [comment]
How about `const AQL = '';`... then we can migrate to `\Sky\Model`.

i.e. `\Vendor\Model\my_model.php`

`$aql = my_model::AQL;`

**Comment by will123195** [comment]
Here's my idea to clean up the models a bit

``` php
<?php

class album extends \Sky\Model {

    const AQL = '
        album {
            title,
            year
        }';

    protected $_config = [
        'belongs_to' => [], // list of models this is a submodel of ($_belongs_to)
        'mod_time' => null, // discard cache data older than this date/time
        'possible_errors' => [], // error codes and corresponding error messages
        'required_fields' => [], // list of required fields
        'read_only' => [], // do not save these fields, tables, subs (nested tables), or objects
                           // ($_ignore)... do we need "subs"?
        'test_mode' => false, // will not save anything if true ($_abort_save)
    ];

/*
    // internal data that is generated at runtime, we could cache this
    public $_metadata = [
        'errors' => 
        'primary_table' => 
        'aql_array' => 
        ...
    ];

    // instance data
    protected $_data = [
        'id' =>
        'ide' =>
        'title' => 
        'year' => 
        'additional_property' =>
    ];
*/

    // maybe these should be abstract
    // do we really need to be able to instantiate Model?
    public function beforeCheckRequiredFields() {}
    public function validate() {}
    private function beforeInsert() {}
    private function beforeUpdate() {}
    private function beforeRefreshCache() {}
    private function afterInsert() {}
    private function afterUpdate() {}

}
```

Other topics for discussion:
- can we get rid of `addProperty()`
- `$_data` should be cached after both `loadDB()` and `construct()` so additional properties can be cached

**Comment by stanistan** [comment]
Additional properties don't make sense in data, Data should specifically be model data that were caching. The cache object should have keys: data, cache_time. 

Maybe there should be two hooks, afterLoad and construct, the stuff in after load gets cached.

**Comment by will123195** [comment]
I agree `$_data` should only contain model data.. I removed `cached_time` from `$_data`.  However, `addProperty()` is used very frequently and those "calculated" fields need to be cached and it would be redundant to put them in the database.

Are you saying you want to always run construct() even if the data is cached?

**Comment by stanistan** [comment]
Not necessarily, I was just thinking about in the current invocation of `construct()`, and not only running after a real db load.

---

I agree, `Model` should probably be abstract.

---

In the previous version of this, we have the ability for `aql` to return `Model` objects, I don't think this is good design anymore, the two of these classes should **not** be this tightly coupled together.

**Comment by tolijoonbug** [comment]
My opinions:

Let's get rid of model set_ methods.
- Can we have validateField methods. I like that it "forces" the user to separate logic for validating each field. Then you can just see in an IDE what is being validated and what's not.

I think keep the .aql separate from the model is a good thing. Writing long multiline strings in PHP is annoying. It's kind of the reason for keeping HTML out of php files.

belongs_to should be deprecated. Instead any time a Model loads it should build it automatically based on other loaded classes. Would prevent programmer errors and take minimal resources to do automatically

Cache should NOT include properties. It will cause a lot of weird ambiguities in how to use Properties. Ex: I would do if ($password1) { .... }. If it's already in the cache and I never entered it it fill give a false positive. cache should match DB.

**Comment by stanistan** [comment]
### Set methods:

Changing my mind, we should keep `set_` methods, definitely change the name to something else, `validate_` maybe.

If we don't have these field specific methods, than for every time you want to validate a field the dev would have to rewrite the same logic.

```
// pseudo code:
if (fieldIsSet(name) && fieldDoesNotHaveErrors(name)) {
    // do some stuff here  
}
```

And for this, it would be best practice to define those as methods regardless (so they're modular), and we wouldn't have a lot of redundant logic.

---

### Model::AQL

These are the benefits/drawbacks of both approaches that I see to keeping the AQL separate or together with the class declaration.

##### For AQL in class definition:
- one file to maintain
- only one file to read (optimization, but probably minor)
- we can remove the model specific autoloader (no `model_name` directories)

##### For `.aql` files
- separation of data structure from validation object
- cleaner class files, as some models are large
- AQL _is_ a different language, so keeping it separate from PHP is a good idea

---

**Comment by will123195** [comment]
resolved in 3.0
